File: 02/python/02.py
#!/usr/bin/env python3
from collections import Counter
from itertools import combinations

import numpy as np


if __name__ == '__main__':
    with open('../02.txt', 'r') as f:
        l = [s.strip() for s in f.readlines()]

    counts = [0, 0]
    for s in l:
        c = Counter(s)
        counts[0] += 2 in c.values()
        counts[1] += 3 in c.values()
    chksum = counts[0] * counts[1]

    # Output part 1 result
    print(chksum)

    # Find strings with smallest Levenshtein distance
    l2n = {c: i for i, c in enumerate('abcdefghijklmnopqrstuvwxyz')}

    combs = np.array(list(combinations(range(len(l)), 2)))
    pairs = np.array([list(map(l2n.get, s)) for s in l])[combs]
    diff = pairs[:,1,:] - pairs[:,0,:]
    min_idx = np.argmin(np.count_nonzero(diff, axis=1))

    similar = ''.join(
        np.array(list(l[combs[min_idx, 0]]))[diff[min_idx] == 0]
    )

    # Output part 2 result
    print(similar)

File: 02/go/02.go
package main

import (
	"fmt"
	"io/ioutil"
	"strings"
)

func main() {
	// Load data from file
	data, err := ioutil.ReadFile("../02.txt")
	if err != nil {
		panic(err)
	}
	lines := strings.Split(string(data), "\n")
	lines = lines[:len(lines)-1] // drop empty line at end

	// Get chksum for list of ids
	counts := []int{0, 0}

	for _, s := range lines {
		m := make(map[rune]int)
		for _, c := range s {
			m[c]++
		}
		var twos, threes int
		for _, v := range m {
			switch v {
			case 2:
				twos = 1
			case 3:
				threes = 1
			}
			if twos == 1 && threes == 1 {
				break
			}
		}
		counts[0] += twos
		counts[1] += threes
	}

	chksum := counts[0] * counts[1]

	// Output part 1 result
	fmt.Println(chksum)

	// Find strings with smallest Levenshtein distance
	min_dist := len(lines[0])
	min_i, min_j := len(lines), len(lines)
	for i, s1 := range lines {
		for j, s2 := range lines[:i] {
			dist := 0
			for k := 0; k < len(s1); k++ {
				if s1[k] != s2[k] {
					dist++
					if dist == min_dist {
						break
					}
				}
			}
			if dist < min_dist {
				min_dist, min_i, min_j = dist, i, j
			}
		}
	}

	similar := make([]byte, 0, len(lines[0]))
	s1, s2 := lines[min_i], lines[min_j]
	for i := 0; i < len(s1); i++ {
		if s1[i] == s2[i] {
			similar = append(similar, s1[i])
		}
	}

	// Output part 2 result
	fmt.Println(string(similar))
}
